include/revng/TupleTree/TrackingVisits.h

```diff
@@ -0,0 +1,239 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+#include <tuple>
+
+#include "revng/ADT/KeyedObjectContainer.h"
+#include "revng/ADT/TrackingContainer.h"
+#include "revng/ADT/UpcastablePointer.h"
+#include "revng/Support/YAMLTraits.h"
+#include "revng/TupleTree/TupleLikeTraits.h"
+#include "revng/TupleTree/TupleTreeCompatible.h"
+#include "revng/TupleTree/TupleTreePath.h"
+#include "revng/TupleTree/Visits.h"
+
+struct ReadFields {
+  std::set<TupleTreePath> Read;
+  std::set<TupleTreePath> ExactVectors;
+};
+
+namespace revng::detail {
+
+template<typename M>
+struct ReadFieldsCollector {
+  TupleTreePath Stack;
+
+  ReadFields Info;
```

Comment:
Promote to local variable of `collect` and pass them down as arguments to `collectImpl`s.